### PR TITLE
Test framework

### PR DESCRIPTION
**Pull request by pmelab**

### CODE REVIEW

File: tests/src/Traits/MockSchemaTrait.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+  public function schemaManagerFactory() {
+    return $this->schemaManagerProphecy->reveal();
+  }
+
+  protected function registerSchemaPluginManager(ContainerBuilder $container) {
+    if ($this instanceof KernelTestBase) {
+      $this->schemaManagerProphecy = $this->prophesize(SchemaPluginManager::class);
+
+      $that = $this;
```
Comment (fubhy) [inline tests/src/Traits/MockSchemaTrait.php]: Why is this required?

File: tests/src/Traits/MockSchemaTrait.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+   * @var \Prophecy\Prophecy\ObjectProphecy
+   */
+  protected $schemaManagerProphecy;
+
+  public function schemaManagerFactory() {
+    return $this->schemaManagerProphecy->reveal();
+  }
+
+  protected function registerSchemaPluginManager(ContainerBuilder $container) {
+    if ($this instanceof KernelTestBase) {
```
Comment (fubhy) [inline tests/src/Traits/MockSchemaTrait.php]: Why do we need this check?

File: tests/src/Traits/MockTypeSystemTrait.php [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+   */
+  protected $typeSystemClassMap = [];
+
+  public function typeSystemPluginManagerFactory($id) {
+    return $this->typeSystemPluginManagerProphecies[$id]->reveal();
+  }
+
+  protected function registerTypeSystemPluginManagers(ContainerBuilder $container) {
+    if ($this instanceof KernelTestBase) {
+      $that = $this;
```
Comment (fubhy) [inline tests/src/Traits/MockTypeSystemTrait.php]: Same here? Why is this needed?

File: tests/src/Traits/MockSchemaTrait.php (hunk 1 above)
Comment (pmelab) [reply tests/src/Traits/MockSchemaTrait.php]: The prophecy callback is bound to the prophecy object, so `$this` won't work. But perhaps there's a more elegant way around this.

File: tests/src/Traits/MockSchemaTrait.php (hunk 2 above)
Comment (pmelab) [reply tests/src/Traits/MockSchemaTrait.php]: Because somebody might attach this to something else than a KernelTest. And PHPStorm gives me autocompletion this way 😛

File: tests/src/Traits/MockTypeSystemTrait.php (hunk 3 above)
Comment (pmelab) [reply tests/src/Traits/MockTypeSystemTrait.php]: see above

File: tests/src/Traits/MockSchemaTrait.php (hunk 2 above)
Comment (fubhy) [reply tests/src/Traits/MockSchemaTrait.php]: Right but then just use the /** typehint (via comments)?

File: tests/src/Traits/MockSchemaTrait.php (hunk 1 above)
Comment (fubhy) [reply tests/src/Traits/MockSchemaTrait.php]: Interesting. Do they somehow overwrite the normal context (which is "$this from the place you create the closure in") through reflection?

Comment (pmelab) [reply tests/src/Traits/MockSchemaTrait.php]: It uses `Closure::bind`. We could just introduce our own callback promise to avoid that.

**Comment by pmelab** [comment]
@fubhy Finally done. The base module tests are all ported too.

**Comment by fubhy** [comment]
This looks AMAZING @pmelab

**Comment by wimleers** [comment]
Nice work!

@pmelab specifically pointed me to `tests/src/Kernel/GraphQLTestBase.php`, which looks nice and simple. I do have two questions about that class:

1. Why a kernel test rather than an integration test? It's faster, but it also means you're not testing with actual requests like the client would make them?
2. Why do responses vary by the `user` cache context by default? Why not `user.permissions`?
```
  protected function defaultCacheContexts() {
    return ['gql', 'languages:language_interface', 'user'];
  }
```

**Comment by pmelab** [comment]
@wimleers:

1. For performance and focus. Parts of the tests issue requests to the HttpKernel, to test the cache subscriber. But the main test interface for extensions is the Processor. They should not bother with the framework.
Nothing stops us running am set of webtests using a mocked schema though.

2. Yep, also discovered this while writing tests 😀
